### ISSUE TITLE
when I use model.generate (), the gist_mask metrix seems don't work

### ISSUE THREAD

**Issue by pengfeiwu1999** [jayelm/gisting#14]
In the 180 line of the generation_utils_gist.py file, the new gist attention mask seems to be all one !, so when I use model.generate() function, the generated token  can see all the previous token, My understanding is not necessarily correct, but also hope to clarify the confusion， thanks

**Comment by jayelm** [comment]
Hi,

Which model are you using? There are two gist masks being passed through: `attention_mask` and `attention_mask_gist`. Are you sure you're looking at the right one? Typically `attention_mask` is the vanilla attention mask created by HF, while `attention_mask_gist` are the gist limitations . The full attention mask is created by ANDing `attention_mask`, `attention_mask_gist`, and (when relevant) the decoder autoregressive attention mask, e.g. https://github.com/jayelm/gisting/blob/main/src/gist_llama.py#L529-L542

**Comment by pengfeiwu1999** [comment]
> I'm sure I look the attention_mask_gist, in the 180 line of the generation_utils_gist.py file, I found there's a bug, when your input format is "Prompt <GIST>", the first attention_mask_gist is all 1 matrix, so in your code, the generate progress will continue to generate the "attention_mask_gist" as all 1 matrix, could you please check whether I'm right? thanks!

**Comment by jayelm** [comment]
Could you clarify which model you're using, and the exact code you use to decode from the model, so I can reproduce this?

**Comment by jayelm** [comment]
To clarify these lines specifically:

https://github.com/jayelm/gisting/blob/main/src/generation_utils.py#L177-L187

The reason why we add all ones here is because we are extending the existing gist mask to account for the token we just generated. (This function, `_update_model_kwargs_for_generation`, is called after sampling a new token). Future tokens should be able to attend to the token we just sampled, hence why we add all 1s here.

Notice that if your input is literally just `"Prompt "` and there are no <GIST> tokens already in your input, then you're right that the gist attention mask will always be ones, since you haven't actually added a gist token in your input.

In the case where attention mask gist *does* start out with some masking applied, this update function preserves the masking already present by taking the last row of the gist mask (which might have some 0s to prevent tokens from attending pre-gist) and appending a 1 to the end (since again, the new token we just sampled should be attended to in subsequent generations). Notice that we do not completely replace the old gist attention mask—we are adding to the end. So any existing gist masking is preserved.

You can read the docstring of the function we're looking at for additional context.

Let me know if that answers your question—if not please send me a specific code snippet you use to reproduce the results and I'll take a closer look.

**Comment by jayelm** [comment]
Closing this but let me know if your concerns are unresolved!
